internal/urlutil/urlutil.go

```diff
@@ -71,3 +72,20 @@ func ExtractHostname(addr string) (string, error) {
 	}
 	return u.Hostname(), nil
 }
+
+// From: https://golang.org/src/net/http/transport.go:2712
+var portMap = map[string]string{
+	"http":  "80",
+	"https": "443",
+}
+
+// CanonicalAddr returns url.Host but always with a ":port" suffix
+// From: https://golang.org/src/net/http/transport.go:2719 without idna to ascii conversion
+func CanonicalAddr(url *url.URL) string {
+	addr := url.Hostname()
+	port := url.Port()
+	if port == "" {
+		port = portMap[url.Scheme]
```

Comment:
how does this interact with [downloader plugins](https://helm.sh/docs/topics/plugins/#downloader-plugins) that can specify an alternative URL scheme, like `git://` or `oci://`? Looks like this only handles more "traditional" URL schemes but cannot account for other getter schemes registered with the helm client.

Can you add a test for those?

Reply:
I've tried my hand at this, adding two test cases:
```
		{
			Scheme: "myprotocol",
			Host:   "example.com:1234",
			Path:   "/",
		}: "example.com:1234",
```
This test fails, because the existing code would append a `:` even if looking up a port number from the map comes up empty.

I also have a fix for that (very little experience with go, so probably not ideal). To share this code, should I open a PR against the branch in the fork by @dol or what would be proper Github etiquette for this?

Reply:
@bacongobbler Thank you for the hint. I wasn't aware of this plugin capability of helm. I need to do some digging to understand the topic in more depth.
At the moment I don't see any reason to add other protocols and their default port mapping. I currently considering to change the function name to `HttpCanonicalAddr` to give a clear hint that the function is only meant to be used for http schemes.

Reply:
I am unclear: Is there more to do on this PR, or is this for later/future work?

Reply:
If the CanonicalAddr function is used in cases of downloader plugins, IMHO the PR must ensure that URL schemes that are not covered by the mapping are not changed. Currently, `myprotocol://example.com` would get "canonicalized" to `myprotocol://example.com:` which is probably not good.

I've pushed a rudimentary "fix" here: https://github.com/dol/helm/pull/52

Reply:
IMHO there is nothing additional to do.
The mentioned [Downloader Plugins](https://helm.sh/docs/topics/plugins/#downloader-plugins) allows for other protocols. The custom protocol handler is an external command that's outside of helm. For this reason the focus on http as protocol is IMHO enough.

Reply:
Okay, that partially makes sense.

However what I was referring to was the use of CanonicalAddr in the `action` package on install.go, line 693. It is entirely possible for that line of code to receive a URL like the test case @erpel provided above, as u1 could be a plugin getter.

In other words, if you intend to use this ONLY in the context of HTTP/HTTPS protocols, then it should be used ONLY in httpgetter.go.

Hope that makes sense.